### PR TITLE
add AMOP authority support

### PR DESCRIPTION
**Pull request by sulenn**

### CODE REVIEW

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  2
- examples/pubauth/publisher.go  3
- examples/subauth/subscriber.go  3
- examples/subscriber/subscriber.go  3
- examples/publisher_for_multi_broadcast/publisher.go  5
- examples/publisher_for_unique_broadcast/publisher.go  5
         

Clones added
============
- conn/channel.go  13
- examples/pubauth/publisher.go  3
- examples/subauth/subscriber.go  2
- examples/subscriber/subscriber.go  2
- examples/publisher_for_multi_broadcast/publisher.go  4
- examples/publisher_for_unique_broadcast/publisher.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  2
- examples/publisher_for_unique_broadcast/publisher.go  5
- examples/publisher_for_multi_broadcast/publisher.go  5
- examples/subscriber/subscriber.go  3
- examples/subauth/subscriber.go  3
- examples/pubauth/publisher.go  3
         

Clones added
============
- conn/channel.go  13
- examples/publisher_for_unique_broadcast/publisher.go  4
- examples/publisher_for_multi_broadcast/publisher.go  4
- examples/subscriber/subscriber.go  2
- examples/subauth/subscriber.go  2
- examples/pubauth/publisher.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 19
           

Complexity increasing per file
==============================
- examples/amop/sub/subscriber.go  4
- examples/amop/unicast_pub/publisher.go  5
- examples/amop_auth/sub/subscriber.go  5
- examples/amop_auth/multicast_pub/publisher.go  7
- conn/channel.go  4
- examples/amop/multicast_pub/publisher.go  5
- examples/amop_auth/unicast_pub/publisher.go  7
         

Clones added
============
- examples/amop/sub/subscriber.go  2
- examples/amop/unicast_pub/publisher.go  5
- examples/amop_auth/sub/subscriber.go  1
- examples/amop_auth/multicast_pub/publisher.go  5
- examples/amop/multicast_pub/publisher.go  5
- examples/amop_auth/unicast_pub/publisher.go  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

File: examples/README.md [hunk 1]
```diff
@@ -2,52 +2,132 @@
...
+  publish topic success
+  2020/08/10 23:18:50 publish message: Hi, FISCO BCOS! 0 
+  2020/08/10 23:18:52 PushAuthTopicDataRandom failed, err: sendMessage failed, err: error code 100, remote peer unavailable
+  2020/08/10 23:18:52 publish message: Hi, FISCO BCOS! 1 
+  2020/08/10 23:18:54 publish message: Hi, FISCO BCOS! 2 
+  2020/08/10 23:18:56 publish message: Hi, FISCO BCOS! 3 
+  2020/08/10 23:18:58 publish message: Hi, FISCO BCOS! 4
+  ```
+
+- 启动 AMOP 消息订阅者：go run examples/amop_auth/sub/subscriber.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:97]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 2]
```diff
@@ -2,52 +2,132 @@
...
--   运行 AMOP 客户端：
-
-    ```shell
-    go run examples/singleclient/unicast_client.go
-    ```
-
-    服务端控制台输出 "hello, multi broadcasts!"，客户端输出 "PushTopicDataToALL success" 表示运行成功
+**多播** 指的时节点向监听同一个 Topic 的所有订阅者转发消息。只要网络正常，即使没有监听 Topic 的订阅者，消息发布者也会收到消息发送成功的网络回包
+
+- 运行 AMOP 消息发布者：go run examples/amop/multicast_pub/publisher.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:55]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 3]
```diff
@@ -2,52 +2,132 @@
...
+  ```
 
--   运行 AMOP 客户端：
 
-    ```shell
-    go run examples/singleclient/unicast_client.go
-    ```
-
-    服务端控制台输出 "hello, unique broadcast!"，客户端输出 "PushTopicDataRandom success" 表示运行成功
+> 注意：
```
Comment (fisco-dev) [inline examples/README.md:47]: Codacy found an issue: [[no-consecutive-blank-lines] Remove 1 line before node](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 4]
```diff
@@ -2,52 +2,132 @@
...
 
--   实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
--   可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
--   高效：AMOP消息结构简洁、处理逻辑高效，仅需少量cpu占用，能充分利用网络带宽。
--   安全：AMOP的所有通讯链路使用SSL加密，加密算法可配置,支持身份认证机制。
--   易用：使用AMOP时，无需在SDK做任何额外配置。
+- 实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
+- 可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
+- 高效：AMOP消息结构简洁、处理逻辑高效，仅需少量cpu占用，能充分利用网络带宽。
+- 安全：AMOP的所有通讯链路使用SSL加密，加密算法可配置,支持身份认证机制。
```
Comment (fisco-dev) [inline examples/README.md:8]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 5]
```diff
@@ -2,52 +2,132 @@
...
+  Subscriber success
+  2020/08/10 23:18:52 received: Hi, FISCO BCOS! 1
+  2020/08/10 23:18:54 received: Hi, FISCO BCOS! 2
+  2020/08/10 23:18:56 received: Hi, FISCO BCOS! 3
+  2020/08/10 23:18:58 received: Hi, FISCO BCOS! 4
+  ```
+
+## 身份验证多播案例
+
+- 运行 AMOP 消息发布者：go run examples/amop_auth/multicast_pub/publisher.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:111]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 6]
```diff
@@ -2,52 +2,132 @@
...
--   搭建单群组四节点区块链网络，可参考：[安装](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/installation.html)。
+- 搭建单群组四节点区块链网络，可参考：[安装](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/installation.html)。
 
 ## 单播案例
 
-**单播** 就是节点从监听同一 Topic 的多个订阅者中随机抽取一个订阅者转发消息
+**单播** 指的是节点从监听同一 Topic 的多个订阅者中随机抽取一个订阅者转发消息
 
--   启动 AMOP 服务端：
+- 运行 AMOP 消息发布者：go run examples/amop/unicast_pub/publisher.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:21]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 7]
```diff
@@ -2,52 +2,132 @@
...
+  2020/08/10 22:54:12 received: hello, FISCO BCOS, I am multi broadcast publisher! 3
+  2020/08/10 22:54:14 received: hello, FISCO BCOS, I am multi broadcast publisher! 4
+  2020/08/10 22:54:16 received: hello, FISCO BCOS, I am multi broadcast publisher! 5
+  ```
+
+## 身份验证单播案例
+
+在普通的配置下，任何一个监听了某topic的接收者都能接受到发送者推送的消息。但在某些场景下，发送者只希望特定的接收者能接收到消息，不希望无关的接收者能任意的监听此topic。在此场景下，需要使用Topic认证功能。 认证功能是指对于特定的topic消息，允许通过认证的接收者接收消息。详细请参考：[Topic认证功能](https://fisco-bcos-doc-qiubing.readthedocs.io/en/latest/docs/manual/amop_protocol.html#topic)
+
+- 运行 AMOP 消息发布者：go run examples/amop_auth/unicast_pub/publisher.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:83]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 8]
```diff
@@ -2,52 +2,132 @@
 
 AMOP（Advanced Messages Onchain Protocol）即链上信使协议，旨在为联盟链提供一个安全高效的消息信道，联盟链中的各个机构，只要部署了区块链节点，无论是共识节点还是观察节点，均可使用AMOP进行通讯，AMOP有如下优势：
 
--   实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
--   可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
--   高效：AMOP消息结构简洁、处理逻辑高效，仅需少量cpu占用，能充分利用网络带宽。
--   安全：AMOP的所有通讯链路使用SSL加密，加密算法可配置,支持身份认证机制。
--   易用：使用AMOP时，无需在SDK做任何额外配置。
+- 实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
```
Comment (fisco-dev) [inline examples/README.md:5]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 9]
```diff
@@ -2,52 +2,132 @@
...
+- 可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
+- 高效：AMOP消息结构简洁、处理逻辑高效，仅需少量cpu占用，能充分利用网络带宽。
```
Comment (fisco-dev) [inline examples/README.md:7]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md (hunk 5 above)
Comment (fisco-dev) [inline examples/README.md:111]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 10]
```diff
@@ -2,52 +2,132 @@
...
+- 安全：AMOP的所有通讯链路使用SSL加密，加密算法可配置,支持身份认证机制。
+- 易用：使用AMOP时，无需在SDK做任何额外配置。
 
 进一步了解 AMOP，请参考：[链上信使协议](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/manual/amop_protocol.html)
 
 **初始化**：
 
--   搭建单群组四节点区块链网络，可参考：[安装](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/installation.html)。
+- 搭建单群组四节点区块链网络，可参考：[安装](https://fisco-bcos-documentation.readthedocs.io/zh_CN/latest/docs/installation.html)。
```
Comment (fisco-dev) [inline examples/README.md:15]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 11]
```diff
@@ -2,52 +2,132 @@
 
 AMOP（Advanced Messages Onchain Protocol）即链上信使协议，旨在为联盟链提供一个安全高效的消息信道，联盟链中的各个机构，只要部署了区块链节点，无论是共识节点还是观察节点，均可使用AMOP进行通讯，AMOP有如下优势：
 
--   实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
--   可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
--   高效：AMOP消息结构简洁、处理逻辑高效，仅需少量cpu占用，能充分利用网络带宽。
--   安全：AMOP的所有通讯链路使用SSL加密，加密算法可配置,支持身份认证机制。
--   易用：使用AMOP时，无需在SDK做任何额外配置。
+- 实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
+- 可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
```
Comment (fisco-dev) [inline examples/README.md:6]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 12]
```diff
@@ -2,52 +2,132 @@
...
+    2020/08/10 22:43:33 publish message: hello, FISCO BCOS, I am unique broadcast publisher! 1 
+    2020/08/10 22:43:35 PushTopicDataRandom failed, err: sendMessage failed, err: error code 100, remote peer unavailable
+    2020/08/10 22:43:35 publish message: hello, FISCO BCOS, I am unique broadcast publisher! 2 
+    2020/08/10 22:43:37 PushTopicDataRandom failed, err: sendMessage failed, err: error code 100, remote peer unavailable
+    2020/08/10 22:43:37 publish message: hello, FISCO BCOS, I am unique broadcast publisher! 3 
+    2020/08/10 22:43:39 publish message: hello, FISCO BCOS, I am unique broadcast publisher! 4
+  ```
 
-    控制台输出 "subscribe success" 表示服务启动成功
+- 启动 AMOP 消息订阅者：go run examples/amop/sub/subscriber.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:36]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

Comment (fisco-dev) [inline examples/README.md:36]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md (hunk 2 above)
Comment (fisco-dev) [inline examples/README.md:55]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md (hunk 6 above)
Comment (fisco-dev) [inline examples/README.md:21]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md (hunk 1 above)
Comment (fisco-dev) [inline examples/README.md:97]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md (hunk 7 above)
Comment (fisco-dev) [inline examples/README.md:83]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 13]
```diff
@@ -2,52 +2,132 @@
...
--   实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
--   可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
--   高效：AMOP消息结构简洁、处理逻辑高效，仅需少量cpu占用，能充分利用网络带宽。
--   安全：AMOP的所有通讯链路使用SSL加密，加密算法可配置,支持身份认证机制。
--   易用：使用AMOP时，无需在SDK做任何额外配置。
+- 实时：AMOP消息不依赖区块链交易和共识，消息在节点间实时传输，延时在毫秒级。
+- 可靠：AMOP消息传输时，自动寻找区块链网络中所有可行的链路进行通讯，只要收发双方至少有一个链路可用，消息就保证可达。
+- 高效：AMOP消息结构简洁、处理逻辑高效，仅需少量cpu占用，能充分利用网络带宽。
+- 安全：AMOP的所有通讯链路使用SSL加密，加密算法可配置,支持身份认证机制。
+- 易用：使用AMOP时，无需在SDK做任何额外配置。
```
Comment (fisco-dev) [inline examples/README.md:9]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- conn/channel.go  4
- examples/amop/multicast_pub/publisher.go  5
- examples/amop/sub/subscriber.go  4
- examples/amop/unicast_pub/publisher.go  5
         

Clones added
============
- examples/amop/multicast_pub/publisher.go  3
- examples/amop/sub/subscriber.go  1
- examples/amop/unicast_pub/publisher.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

File: examples/README.md [hunk 14]
```diff
@@ -16,38 +16,59 @@ AMOP（Advanced Messages Onchain Protocol）即链上信使协议，旨在为联
...
+      2020/08/10 22:43:37 publish message: hello, FISCO BCOS, I am unique broadcast publisher! 3 
+      2020/08/10 22:43:39 publish message: hello, FISCO BCOS, I am unique broadcast publisher! 4
     ```
 
-    控制台输出 "subscribe success" 表示服务启动成功
-
--   修改 config.toml 配置文件中 NodeURL 配置项，从 127.0.0.1:20200 修改为 127.0.0.1:20201，因为 FISCO BCOS 2.5.0 不支持 AMOP 连接同一节点（FISCO BCOS 2.6.0 开始支持连接同一节点）
-
--   运行 AMOP 客户端：
+-   启动 AMOP 消息订阅者：go run examples/amop/sub/subscriber.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:36]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 15]
```diff
@@ -16,38 +16,59 @@ AMOP（Advanced Messages Onchain Protocol）即链上信使协议，旨在为联
 
 ## 单播案例
 
-**单播** 就是节点从监听同一 Topic 的多个订阅者中随机抽取一个订阅者转发消息
+**单播** 指的是节点从监听同一 Topic 的多个订阅者中随机抽取一个订阅者转发消息
 
--   启动 AMOP 服务端：
+-   运行 AMOP 消息发布者：go run examples/amop/unicast_pub/publisher.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:21]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

File: examples/README.md [hunk 16]
```diff
@@ -16,38 +16,59 @@ AMOP（Advanced Messages Onchain Protocol）即链上信使协议，旨在为联
...
-    服务端控制台输出 "hello, unique broadcast!"，客户端输出 "PushTopicDataRandom success" 表示运行成功
+> 注意：控制台输出 *PushTopicDataRandom failed, err: sendMessage failed, err: error code 100, remote peer unavailable* 表示网络中没有对应的 topic 消息订阅者
 
 ## 多播案例
 
-**多播** 就是节点向监听同一个 Topic 的所有订阅者转发消息。只要网络正常，即使没有监听 Topic 的订阅者，消息发布者也会收到消息发送成功的网络回包
+**多播** 指的时节点向监听同一个 Topic 的所有订阅者转发消息。只要网络正常，即使没有监听 Topic 的订阅者，消息发布者也会收到消息发送成功的网络回包
 
--   启动 AMOP 服务端：同单播案例
+-   运行 AMOP 消息发布者：go run examples/amop/multicast_pub/publisher.go [endpoint] [topic]
```
Comment (fisco-dev) [inline examples/README.md:52]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  4
- examples/amop/multicast_pub/publisher.go  5
- examples/amop/sub/subscriber.go  4
- examples/amop/unicast_pub/publisher.go  5
         

Clones added
============
- examples/amop/multicast_pub/publisher.go  3
- examples/amop/sub/subscriber.go  1
- examples/amop/unicast_pub/publisher.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- examples/amop/sub/subscriber.go  4
- conn/channel.go  4
- examples/amop/unicast_pub/publisher.go  5
- examples/amop/multicast_pub/publisher.go  5
         

Clones added
============
- examples/amop/sub/subscriber.go  1
- examples/amop/unicast_pub/publisher.go  3
- examples/amop/multicast_pub/publisher.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  4
- examples/amop/sub/subscriber.go  4
- examples/amop/multicast_pub/publisher.go  5
- examples/amop/unicast_pub/publisher.go  5
         

Clones added
============
- examples/amop/sub/subscriber.go  1
- examples/amop/multicast_pub/publisher.go  3
- examples/amop/unicast_pub/publisher.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

File: conn/channel.go [hunk 17]
```diff
@@ -149,9 +180,14 @@ func (t *topicData) Encode() []byte {
 	if err != nil {
 		log.Fatal("encode type error:", err)
 	}
+
+	// err = binary.Write(buf, binary.BigEndian, uint16(len(t.data)))
```
Comment (bxq2011hust) [inline conn/channel.go:184]: delete this

File: conn/channel.go [hunk 18]
```diff
@@ -234,6 +270,11 @@ func decodeTopic(raw []byte) (*topicData, error) {
 		fmt.Println("binary.Read failed:", err)
 	}
 	result.topic = string(topic)
+	// var dataLength uint16
+	// err = binary.Read(buf, binary.BigEndian, &dataLength)
```
Comment (bxq2011hust) [inline conn/channel.go:274]: delete this

File: conn/channel.go [hunk 19]
```diff
@@ -543,26 +600,206 @@ func (hc *channelSession) subscribeTopic(topic string, handler func([]byte)) err
 	if _, ok := hc.topicHandlers[topic]; ok {
 		return errors.New("already subscribed to topic " + topic)
 	}
+	if _, ok := hc.topicHandlers[needVerifyPrefix+topic]; ok {
```
Comment (bxq2011hust) [inline conn/channel.go:603]: delete this

File: conn/channel.go [hunk 20]
```diff
@@ -543,26 +600,206 @@ func (hc *channelSession) subscribeTopic(topic string, handler func([]byte)) err
...
 	hc.topicMu.Unlock()
 
-	keys := make([]string, 0, len(hc.topicHandlers))
-	for k := range hc.topicHandlers {
-		keys = append(keys, k)
+	return hc.sendSubscribedTopics()
+}
+
+func (hc *channelSession) publishAuthTopic(topic string, publicKeys []*ecdsa.PublicKey, handler func([]byte)) error {
+	if len(topic) > maxTopicLength {
```
Comment (bxq2011hust) [inline conn/channel.go:614]: should len(topic) + perfix + random <= 254

File: conn/channel.go [hunk 21]
```diff
@@ -543,26 +600,206 @@ func (hc *channelSession) subscribeTopic(topic string, handler func([]byte)) err
...
-	if err != nil {
-		hc.topicMu.Lock()
-		delete(hc.topicHandlers, topic)
-		hc.topicMu.Unlock()
-		return errors.New("marshal topics failed")
+	if handler == nil {
+		return errors.New("handler is nil")
 	}
-	msg, err := newChannelMessage(amopSubscribeTopics, data)
+	if _, ok := hc.topicHandlers[topic]; ok {
```
Comment (bxq2011hust) [inline conn/channel.go:620]: delete this

File: conn/channel.go [hunk 22]
```diff
@@ -543,26 +600,206 @@ func (hc *channelSession) subscribeTopic(topic string, handler func([]byte)) err
...
+		}
+		hc.messageMu.Unlock()
+	}
+	hc.topicMu.Unlock()
+
+	return hc.sendSubscribedTopics()
+}
+
+func (hc *channelSession) subscribeAuthTopic(topic string, privateKey *ecdsa.PrivateKey, handler func([]byte)) error {
+	if len(needVerifyPrefix+topic) > maxTopicLength {
```
Comment (bxq2011hust) [inline conn/channel.go:715]: should len(topic) + perfix + random <= 254

File: conn/channel.go [hunk 23]
```diff
@@ -543,26 +600,206 @@ func (hc *channelSession) subscribeTopic(topic string, handler func([]byte)) err
...
+		return errors.New("the length of auth topic exceeds 254, because of prefix \"#!$TopicNeedVerify_\"")
+	}
+	if handler == nil {
+		return errors.New("handler is nil")
+	}
+	if _, ok := hc.topicHandlers[topic]; ok {
```
Comment (bxq2011hust) [inline conn/channel.go:721]: delete this

File: conn/channel.go [hunk 24]
```diff
@@ -543,26 +600,206 @@ func (hc *channelSession) subscribeTopic(topic string, handler func([]byte)) err
...
+	for k := range hc.topicHandlers {
+		if strings.Contains(k, authChannelTopicName) {
+			authChannelTopicName = k
+			break
+		}
+	}
+	hc.topicMu.Lock()
+	delete(hc.topicHandlers, authTopicName)
+	delete(hc.topicHandlers, authChannelTopicName)
+	delete(hc.topicHandlers, pushChannelPrefix+authTopicName)
```
Comment (bxq2011hust) [inline conn/channel.go:799]: good

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- examples/amop/unicast_pub/publisher.go  5
- conn/channel.go  4
- examples/amop/multicast_pub/publisher.go  5
- examples/amop/sub/subscriber.go  4
         

Clones added
============
- examples/amop/unicast_pub/publisher.go  3
- examples/amop/multicast_pub/publisher.go  3
- examples/amop/sub/subscriber.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  4
- examples/amop/unicast_pub/publisher.go  5
- examples/amop/sub/subscriber.go  4
- examples/amop/multicast_pub/publisher.go  5
         

Clones added
============
- examples/amop/unicast_pub/publisher.go  3
- examples/amop/sub/subscriber.go  1
- examples/amop/multicast_pub/publisher.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

**Comment by fisco-dev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- conn/channel.go  4
- examples/amop/sub/subscriber.go  4
- examples/amop/unicast_pub/publisher.go  5
- examples/amop/multicast_pub/publisher.go  5
         

Clones added
============
- examples/amop/sub/subscriber.go  1
- examples/amop/unicast_pub/publisher.go  3
- examples/amop/multicast_pub/publisher.go  3
         

Clones removed
==============
+ conn/channel.go  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/FISCO-BCOS/go-sdk/pullRequest?prid=5979849&bid=19608580)

**Pull request merged**